Trigger workflow input usage docs (#46)
slackapi/deno-slack-api: docs/triggers/trigger-generic-inputs.md

# Trigger Generics

When creating or updating a trigger at runtime, a Workflow definition can be passed into the `client.workflows.triggers.create` and `client.workflows.triggers.update` methods as a generic which will allow for typeahead/autocomplete of the input parameters based on the definition's inputs. 

## Defining A Workflow

Recall that in the [SDK](https://github.com/slackapi/deno-slack-sdk/blob/main/docs/workflows.md#workflows), a `workflow` can be defined using the `DefineWorkflow` method.

> When a workflow is created in this manner, if contains a `workflow.definition` object. By passing this definition into the `create` method, the trigger will have access to information related to the workflow's definition.

Reviewer: ```suggestion
When a workflow is created in this manner, it contains a `workflow.definition` object. By passing this definition into the `create` method, the trigger will have access to information related to the workflow's definition.
```

### Using Trigger Generics at runtime Example 

```ts

> import type { SlackFunction } from "/deno-slack-sdk/src/types.ts";

Reviewer: ```suggestion
import { SlackFunction } from "deno-slack-sdk/mod.ts";
```

import { TriggersFunction, TriggersWorkflow } from "../manifest.ts";

> import { SlackAPI } from "../../deno-slack-api/src/mod.ts";

Reviewer: ```suggestion
import { SlackAPI } from "deno-slack-api/mod.ts";
```

> const triggers: SlackFunction<typeof TriggersFunction.definition> = 

Reviewer: You should follow the lead on this doc on how to fill out function runtime logic: https://api.slack.com/future/functions/custom#implement

Ensure you test it out yourself to make sure it works!

Reviewer: @jmplexi you don't need to pass in the generic here anymore. Check the link I linked to above, instead the workflow definition is passed in as the first parameter to `SlackFunction`.

# Trigger Generics

> When creating or updating a trigger at runtime, a Workflow definition can be passed into the `client.workflows.triggers.create` and `client.workflows.triggers.update` methods as a generic which will allow for typeahead/autocomplete of the input parameters based on the definition's inputs. 

Reviewer 2: ```suggestion
When creating or updating a trigger, a Workflow definition can be passed into the `client.workflows.triggers.create` and `client.workflows.triggers.update` methods as a generic which will allow for typeahead/autocomplete of the input parameters based on the definition's inputs. 
```

Not just at runtime but CLI too

## Defining A Workflow

Recall that in the [SDK](https://github.com/slackapi/deno-slack-sdk/blob/main/docs/workflows.md#workflows), a `workflow` can be defined using the `DefineWorkflow` method.
When a workflow is created in this manner, it contains a `workflow.definition` object. By passing this definition into the `create` or `update` method, the trigger will have access to information related to the workflow's definition.

> ### Using Trigger Generics at runtime Example 
> 
> ```ts
> import { TriggersWorkflow } from "../manifest.ts";
> import { SlackAPI } from "deno-slack-api/mod.ts";
> 
> const client = SlackAPI(token);
> 
> const triggerReturn = await client.workflows.triggers.create< //Also works for update
>   typeof TriggersWorkflow.definition
> >({
>   type: "webhook",
>   name: "Request Time off",
>   description: "Starts the workflow to request time off",
>   workflow: "#/workflows/reverse_workflow",
>   inputs: { //inputs can now be autofilled from the Workflow Definition
>     a_string: {
>       value: "TEST",
>     },
>     a_channel: {
>       value: "TEST",
>     },
>     b_string: {
>       value: "TEST",
>     },
>   },
> });
> ```

Reviewer 2: ```suggestion
## Workflow Definition Input with Runtime Triggers

When dealing with Triggers at runtime, a Workflow definition input can optionally be passed to the Trigger's `create` or `update` API call as a generic. When a generic is passed, the object argument being defined will have access to typeahead on the `inputs` and `workflow` properties.

### Using Trigger Generics at runtime Example 

```ts
import { TriggersWorkflow } from "../manifest.ts";
import { SlackAPI } from "deno-slack-api/mod.ts";

...
{
  //Inside the function execution logic
  const client = SlackAPI(token);
  
  const triggerReturn = await client.workflows.triggers.create< //Also works for update
    typeof TriggersWorkflow.definition
  >({
    type: "webhook",
    name: "Request Time off",
    description: "Starts the workflow to request time off",
    workflow: "#/workflows/reverse_workflow",
    inputs: { //inputs can now be autofilled from the Workflow Definition
      a_string: {
        value: "TEST",
      },
      a_channel: {
        value: "TEST",
      },
      b_string: {
        value: "TEST",
      },
    },
}
```
```

Felt like runtime triggers could use a brief intro, and gave a little separation between the file imports to indicate we're in function execution logic

> # Trigger Creation Workflow Definition Inputs

Reviewer 2: ```suggestion
# Trigger Generics
```

This isn't specific to Trigger Creation or to the `inputs` property. I think the term "definition input" is confusing here since triggers accept `inputs` and that's not what we should be referring to.

> When creating a trigger at runtime, a workflow definition can be passed into the `client.workflows.triggers.create` method which will allow for typeahead/autocomplete of the input parameters based on the definition's inputs. 

Reviewer 2: ```suggestion
When creating or updating a Trigger at runtime, a Workflow definition can be passed into the `client.workflows.triggers.create` method as a generic which will allow for typeahead/autocomplete of the input parameters based on the definition's inputs. 
```

Should also mention the `workflow` property

## Defining A Workflow Recap

> Recall in the `manifest.ts` file of an app, a `workflow` can be defined using the `DefineWorkflow` method as follows:

Reviewer 2: ```suggestion
Recall in the `manifest.ts` file of an app, a `Workflow` can be defined using the `DefineWorkflow` method as follows:
```

```ts
export const TriggersWorkflow = DefineWorkflow({
  callback_id: "triggers_workflow",
  title: "Triggers Workflow",
  description: "sample runthrough of a trigger creation",
  input_parameters: {
    properties: {
      a_string: {
        type: Schema.types.string,
      },
      b_string: {
        type: Schema.types.string,
      },
      a_channel: {
        type: Schema.slack.types.channel_id,
      },
    },
    required: [],
  },
});
```

> When a workflow is created in this manner, if contains a `workflow.definition` object. By passing this definition into the `create` method, the trigger will have access to information related to the workflow's definition.

Reviewer 2: We shouldn't add documentation for how a SDK feature works in this repo as it will very likely become out of date. Link out to the SDK docs or to the public API docs instead.

The only exception is if we're actually using the values from this workflow in the examples below, which would mean that the examples should provide matching `input_parameters` based on the `required` field and that the value passed to `workflow` matches.

> ### Using Workflow definition input at runtime Example 

Reviewer 2: ```suggestion
### Using Trigger Generics at Runtime Example
```

```ts
import type { SlackFunctionHandler } from "../../deno-slack-sdk/src/types.ts";
import { TriggersFunction, TriggersWorkflow } from "../manifest.ts";
import { SlackAPI } from "../../deno-slack-api/src/mod.ts";

> const triggers: SlackFunctionHandler<typeof TriggersFunction.definition> = 

Reviewer 2: We should use `SlackFunction` in examples rather than `SlackFunctionHandler`

  async (
    { token, env },
  ) => {
    const client = SlackAPI(token, {
      slackApiUrl: env["SLACK_API_URL"],
    });

>     const workflowDef = TriggersWorkflow.definition;
>     const triggerResponse = await client.workflows.triggers.create<
>       typeof workflowDef //Definition passed in as a generic
>     >({

Reviewer 2: ```suggestion
    const triggerResponse = await client.workflows.triggers.create<
      typeof workflowDef.definition //Definition passed in as a generic
    >({
```

Is there a reason we'd recommend them to instantiate a variable for the workflow definition rather than doing it directly in the generic? My two cents is that accessing the definition directly in the generic should be the recommended way
